### PR TITLE
Module 3 - First basic logistic regression

### PR DESCRIPTION
**Pull request by carleondel**
I tried to create the mvp without looking at the solution from Guille. Will make a second push using the solution in the future.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -10,12 +10,13 @@ python = "^3.11"
 pandas = "^2.0.3"
 matplotlib = "^3.7.2"
 numpy = "^1.25.2"
-sklearn = "^0.0.post7"
```
Comment (pabaldonedo) [inline pyproject.toml:13]: Are not we needing sklrarm anymore?

File: src/module_3/exploration_models.md [hunk 2]
```diff
@@ -0,0 +1,2044 @@
+# Exploration Phase - Linear Models Only - Fast Proof of Concept (POC)
+
+**GOAL: Create a Classification Model that can predict whether or not a person would buy an item we offered them (via push notification) based on behavioral and personal features of that user (user id, ordered before, etc), features of that specific order (date, etc) and features of the items themselves (popularity, price, avg days to buy, etc)**
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:3]: Nice! A line with the conclusiom/actiom irems would be great too

File: src/module_3/exploration_models.md [hunk 3]
```diff
@@ -0,0 +1,2044 @@
...
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import confusion_matrix, classification_report
+from sklearn.metrics import ConfusionMatrixDisplay
+from sklearn.metrics import precision_recall_curve, roc_curve, auc
+```
+
+
+```python
+path = r'/home/carleondel/data-zrive-ds/box_builder_dataset/feature_frame.csv'
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:32]: Tip: have a lookt at Path object from pathlib package

File: src/module_3/exploration_models.md [hunk 4]
```diff
@@ -0,0 +1,2044 @@
...
+df = pd.read_csv(path)
+```
+
+## Data
+
+This dataset contains 26 attributes based on the orders users make. Whenever an order is made, we take a picture of the whole inventory. The 'outcome' feature refers to the presence of each inventory's item in the order (0 for no presence, 1 for presence)
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:38]: Awesome! Super helpful for the reader

File: src/module_3/exploration_models.md [hunk 5]
```diff
@@ -0,0 +1,2044 @@
...
+    3442    3643274788996
+    3443    3643283734660
+    3444    3643294515332
+    Name: order_id, Length: 2603, dtype: int64
+
+
+
+
+```python
+# We filter the original df using the filtered order ids
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:747]: Great! Small tip: for things like this that we absolutely know we will be reusing, moving them into functions causes little overhead and we start testing thr method in the notebook, making sure it is bug free

File: src/module_3/exploration_models.md [hunk 6]
```diff
@@ -0,0 +1,2044 @@
...
+```python
+print(f"We have kept {100*len(df_filtered) / len(df):.2f} % of the original dataset")
+```
+
+    We have kept 75.12 % of the original dataset
+
+
+
+```python
+# Quick check to make sure we filtered properly
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:761]: See, this could be presented with almost no effort as a unittest of a method. By doing that you will alreasy have the method and the testimg almost production ready

File: src/module_3/exploration_models.md [hunk 7]
```diff
@@ -0,0 +1,2044 @@
...
+           'people_ex_baby', 'days_since_purchase_variant_id',
+           'avg_days_to_buy_variant_id', 'std_days_to_buy_variant_id',
+           'days_since_purchase_product_type', 'avg_days_to_buy_product_type',
+           'std_days_to_buy_product_type'],
+          dtype='object')
+
+
+
+
+```python
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:796]: Nice practice

File: src/module_3/exploration_models.md [hunk 8]
```diff
@@ -0,0 +1,2044 @@
...
+categorical_cols = ['variant_id', 'product_type', 'order_id', 'user_id', 'vendor']
+binary_cols = ['outcome', 'ordered_before', 'abandoned_before', 'active_snoozed', 'set_as_regular',]
+numerical_cols = [cols for cols in df_filtered.columns if cols not in (categorical_cols+binary_cols)]
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:799]: This has risk if new columns are added. You could list here the numerical colulns and then show if you are missing any columns in the 3 lists

File: src/module_3/exploration_models.md [hunk 9]
```diff
@@ -0,0 +1,2044 @@
...
+# We extract features from our date info
+df_filtered['year'] = df_filtered['created_at'].dt.year
+df_filtered['month'] = df_filtered['created_at'].dt.month
+df_filtered['day'] = df_filtered['created_at'].dt.day
+df_filtered['hour'] = df_filtered['created_at'].dt.hour
+```
+
+    /tmp/ipykernel_2827/1622433905.py:2: SettingWithCopyWarning: 
+    A value is trying to be set on a copy of a slice from a DataFrame.
+    Try using .loc[row_indexer,col_indexer] = value instead
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:861]: Try to follow the warnings

File: src/module_3/exploration_models.md [hunk 10]
```diff
@@ -0,0 +1,2044 @@
...
+    
+    See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
+      df_filtered['created_at'] = pd.to_datetime(df_filtered['created_at'])
+
+
+
+```python
+# We extract features from our date info
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:852]: Interesting. Can you ellaborate what is the rarionale behind incluiding these variables?

File: src/module_3/exploration_models.md [hunk 11]
```diff
@@ -0,0 +1,2044 @@
...
+
+
+
+```python
+# We define the size of each dataset
+train_size = int(0.7 * len(df_filtered))
+val_size = int(0.15 * len(df_filtered))
+
+# We divide the df after being sorted
+df_train = df_filtered.iloc[:train_size]
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:1428]: You can end with an order broken in 2: one part in train and another in test. I suggest to split by the date  that closest gets the desired split ratios

File: src/module_3/exploration_models.md [hunk 12]
```diff
@@ -0,0 +1,2044 @@
...
+df_val = df_filtered.iloc[train_size:train_size + val_size]
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:1429]: Id check how big in days are these sets. They could be a day or a week or a year. And that may be of interest

File: src/module_3/exploration_models.md [hunk 13]
```diff
@@ -0,0 +1,2044 @@
...
+```python
+# Initialize the scaler
+scaler = StandardScaler()
+```
+
+Now we normalize the X train, test and validation data. We only fit to the training data in order to avoid data leakage (statistics coming from the val/test data).
+
+
+```python
+scaled_X_train = scaler.fit_transform(X_train)
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:1460]: Nice to avoid leakage. To make it even better we could use the scaler in a pipeline so we encapsulate the feature processing and predict

File: src/module_3/exploration_models.md [hunk 14]
```diff
@@ -0,0 +1,2044 @@
...
+(Not yet)
+
+
+```python
+
+```
+
+
+```python
+y_pred = model.predict(scaled_X_val)
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:1932]: In a heavily imbalance dataset, the default prediction is discourage. Id expect to see roc or precision recall curves and choosing a threshold

File: src/module_3/exploration_models.md [hunk 15]
```diff
@@ -0,0 +1,2044 @@
...
+             0.0       0.99      1.00      0.99    320198
+             1.0       0.40      0.05      0.09      4394
+    
+        accuracy                           0.99    324592
+       macro avg       0.69      0.53      0.54    324592
+    weighted avg       0.98      0.99      0.98    324592
+    
+
+
+We have a good performance on the negative instances, but a really poor performance for our positive cases. It can be explained by the existing inbalance of our classes in the dataset and also because we need to improve this model.
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:1982]: Without a baseline it is hard to assess the performance, especially not even comparing against prevalence

File: src/module_3/exploration_models.md [hunk 16]
```diff
@@ -0,0 +1,2044 @@
...
+```python
+# We calculate precision and recall
+precision, recall, thresholds = precision_recall_curve(y_val, y_val_proba)
+```
+
+
+```python
+# We plot the precision-recall curve
+
+plt.figure(figsize=(8, 6))
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:2009]: I recommend you get use to call fig, ax = plt.subplots() to be more explicit about the axes you use

File: src/module_3/exploration_models.md [hunk 17]
```diff
@@ -0,0 +1,2044 @@
...
+plt.show()
+```
+
+
+    
+![png](exploration_models_files/exploration_models_54_0.png)
+    
+
+
+We have an AUC of 0.81, much better than guessing at random. But this curve is not really informative in our case since we have very unbalanced classes
```
Comment (pabaldonedo) [inline src/module_3/exploration_models.md:2051]: I am missing choosing the threshold out of these plots, and then evaluating the precision and recall that you chose

**Review by pabaldonedo — APPROVED** [review]
I like how you document the things you do. I'd love a final conclusion though. You make good use of python libraries and have a decent code quality. The general idea of the notebook is correct.

On the other hand, I missed some important points that you should consider in future instances:

- The lack of baseline makes it hard to draw conclusions on performance
- You did not explore different hyperparameters, only ridge
- You did not make any model assessment (how the feature importance look like? for example)
- While you noticed it is an imbalance problem, you do not leverage the precision recall curve for making a decision on the threshold
- You ignored some warnings. If you see warnings, check them.
- You do the split in a way supporting yourself on "temporal data" issues. But there is no plot or analysis why this matters in this dataset.

In addition, Id like to hear from you the justification of taking the day, month and year as features. When making new features or including/excluding some, it is good to argue the motivation.

In general good work but incorporating all of these will make you take a step forward.

Keep up the good work!
